File: src/core/proxy/token-bucket.js
// src/core/proxy/token-bucket.js
//
// Distributed token-bucket rate limiting, Redis-backed. Originally scaffolded
// as an Enterprise-only module (src/ee/rate_limiter) - moved into core once
// it became clear Community-tier rate limiting needed the exact same
// algorithm too (just called via tryConsume() — reject over the cap —
// instead of Enterprise's acquire() — queue and smooth). The tier
// distinction is which method core/proxy calls (gated on ENTERPRISE_MODE),
// not which code exists; src/ee/rate_limiter now just re-exports this for
// anything still requiring that path.
//
// Correctness requires atomicity: reading the current token count and
// then separately writing the decremented value is a race condition
// under concurrent requests - two requests could both read "1 token
// available" and both proceed, over-admitting by exactly the number of
// concurrent racers. A Redis Lua script executes as a single atomic
// operation from Redis's perspective, closing that race entirely.

// Lazy-refill token bucket: rather than running a background timer to
// add tokens, each call computes how many tokens *would* have accrued
// since the last recorded refill time, based on elapsed wall-clock time.
// This is standard practice for distributed token buckets - a timer-based
// refill would need to run somewhere specific (which instance? what if
// it crashes?), while lazy refill is stateless and correct regardless of
// which ar-api instance happens to handle any given request.
const TOKEN_BUCKET_SCRIPT = `
local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4])

local bucket = redis.call('HMGET', key, 'tokens', 'last_refill')
local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])

if tokens == nil then
  tokens = capacity
  last_refill = now
end

local elapsed = math.max(0, now - last_refill)
local refilled = math.min(capacity, tokens + (elapsed * refill_rate))

if refilled >= requested then
  local remaining = refilled - requested
  redis.call('HMSET', key, 'tokens', remaining, 'last_refill', now)
  redis.call('EXPIRE', key, 3600)
  return {1, remaining}
else
  redis.call('HMSET', key, 'tokens', refilled, 'last_refill', now)
  redis.call('EXPIRE', key, 3600)
  -- Time until enough tokens accrue for this request, in seconds.
  local deficit = requested - refilled
  local wait_seconds = deficit / refill_rate
  return {0, wait_seconds}
end
`;

class TokenBucket {
  // redis: an ioredis client instance (reuses the app's existing connection,
  // doesn't open a new one). One TokenBucket instance is meant to be
  // shared across all callers - capacity/refillRate are passed per-call
  // (different identities may have different limits, e.g. free vs pro
  // plans, or a custom per-org override), while the Lua script SHA gets
  // cached once on this instance regardless of what limits get requested.
  constructor(redis) {
    this.redis = redis;
    this.scriptSha = null;
  }

  async _ensureScriptLoaded() {
    if (!this.scriptSha) {
      this.scriptSha = await this.redis.script('load', TOKEN_BUCKET_SCRIPT);
    }
    return this.scriptSha;
  }

  // Attempts to consume `cost` tokens (default 1) from the named bucket.
  // capacity: maximum tokens the bucket can hold (the burst allowance).
  // refillRatePerSecond: steady-state tokens added per second.
  // Returns { allowed: true, remaining } if tokens were available, or
  // { allowed: false, retryAfterSeconds } if not - the caller decides
  // whether to reject immediately or wait and retry (see acquire() below
  // for the waiting variant).
  async tryConsume(bucketKey, capacity, refillRatePerSecond, cost = 1) {
    const sha = await this._ensureScriptLoaded();
    const now = Date.now() / 1000;
    let result;
    try {
      result = await this.redis.evalsha(sha, 1, bucketKey, capacity, refillRatePerSecond, now, cost);
    } catch (err) {
      // NOSCRIPT can happen if Redis restarted and lost cached scripts -
      // reload and retry once rather than failing the request.
      if (err.message && err.message.includes('NOSCRIPT')) {
        this.scriptSha = null;
        const freshSha = await this._ensureScriptLoaded();
        result = await this.redis.evalsha(freshSha, 1, bucketKey, capacity, refillRatePerSecond, now, cost);
      } else {
        throw err;
      }
    }
    const [allowedFlag, value] = result;
    return allowedFlag === 1
      ? { allowed: true, remaining: parseFloat(value) }
      : { allowed: false, retryAfterSeconds: parseFloat(value) };
  }

  // Waits for a token to become available, up to maxWaitMs, rather than
  // rejecting immediately - this is the actual "traffic smoothing"
  // behavior the Enterprise tier adds over Community's reject-on-burst.
  // Returns { acquired: true } once a token is consumed, or
  // { acquired: false, retryAfterSeconds } if the wait would exceed
  // maxWaitMs - an unbounded wait during a genuine downstream outage
  // would just delay the inevitable failure rather than surfacing it, so
  // this is capped rather than infinite.
  async acquire(bucketKey, capacity, refillRatePerSecond, { cost = 1, maxWaitMs = 10000, pollIntervalMs = 100 } = {}) {
    const deadline = Date.now() + maxWaitMs;
    while (true) {
      const result = await this.tryConsume(bucketKey, capacity, refillRatePerSecond, cost);
      if (result.allowed) return { acquired: true, remaining: result.remaining };

      const waitMs = result.retryAfterSeconds * 1000;
      if (Date.now() + waitMs > deadline) {
        return { acquired: false, retryAfterSeconds: result.retryAfterSeconds };
      }
      await new Promise((resolve) => setTimeout(resolve, Math.min(waitMs, pollIntervalMs)));
    }
  }
}

module.exports = { TokenBucket };

File: src/api-gateway-rs/crates/core/src/token_bucket.rs
//! Redis-backed atomic token bucket rate limiter — functionally equivalent
//! to `src/core/proxy/token-bucket.js`'s `TokenBucket` class (lazy refill
//! computed from elapsed wall-clock time, correct across multiple
//! stateless instances, atomic via a Lua script so check-then-act can't
//! race). Not required to be byte-identical to Node's own script — this is
//! ephemeral rate-limit accounting, not data either server needs to read
//! back from the other, unlike the dedup hash format.

use redis::Script;

const SCRIPT: &str = r#"
local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local cost = tonumber(ARGV[3])
local now = tonumber(ARGV[4])

local tokens = capacity
local last_refill = now
local raw = redis.call('GET', key)
if raw then
  local parsed = cjson.decode(raw)
  tokens = parsed.tokens
  last_refill = parsed.last_refill
end

local elapsed = (now - last_refill) / 1000.0
if elapsed > 0 then
  tokens = math.min(capacity, tokens + elapsed * refill_rate)
end

local allowed = 0
if tokens >= cost then
  tokens = tokens - cost
  allowed = 1
end

redis.call('SET', key, cjson.encode({tokens = tokens, last_refill = now}), 'EX', 3600)
return {allowed, tostring(tokens)}
"#;

pub struct ConsumeResult {
    pub allowed: bool,
    pub remaining: f64,
    pub retry_after_seconds: Option<u64>,
}

pub struct TokenBucket {
    script: Script,
}

impl Default for TokenBucket {
    fn default() -> Self {
        Self::new()
    }
}

impl TokenBucket {
    pub fn new() -> Self {
        Self {
            script: Script::new(SCRIPT),
        }
    }

    /// Community-tier behavior: reject immediately over the cap.
    pub async fn try_consume(
        &self,
        conn: &mut redis::aio::MultiplexedConnection,
        bucket_key: &str,
        capacity: f64,
        refill_rate_per_second: f64,
        cost: f64,
    ) -> redis::RedisResult<ConsumeResult> {
        let now_ms = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as i64;

        let (allowed, remaining_str): (i64, String) = self
            .script
            .key(bucket_key)
            .arg(capacity)
            .arg(refill_rate_per_second)
            .arg(cost)
            .arg(now_ms)
            .invoke_async(conn)
            .await?;

        let remaining: f64 = remaining_str.parse().unwrap_or(0.0);
        let retry_after_seconds = if allowed == 0 {
            let deficit = cost - remaining;
            Some((deficit / refill_rate_per_second).ceil().max(1.0) as u64)
        } else {
            None
        };

        Ok(ConsumeResult {
            allowed: allowed == 1,
            remaining,
            retry_after_seconds,
        })
    }

    /// Enterprise-tier behavior: poll until a token frees up or `max_wait`
    /// elapses, instead of rejecting immediately — smooths bursts rather
    /// than failing them. Not wired into any Phase 2 caller yet (Community
    /// mode only so far); included now since it's cheap given try_consume
    /// already exists, and Phase 5 (Enterprise) will need it.
    pub async fn acquire(
        &self,
        conn: &mut redis::aio::MultiplexedConnection,
        bucket_key: &str,
        capacity: f64,
        refill_rate_per_second: f64,
        cost: f64,
        max_wait: std::time::Duration,
    ) -> redis::RedisResult<ConsumeResult> {
        let deadline = std::time::Instant::now() + max_wait;
        loop {
            let result = self
                .try_consume(conn, bucket_key, capacity, refill_rate_per_second, cost)
                .await?;
            if result.allowed || std::time::Instant::now() >= deadline {
                return Ok(result);
            }
            tokio::time::sleep(std::time::Duration::from_millis(100)).await;
        }
    }
}
